// @aws-sdk/client-cloudcontrol@3.1069.0/dist-cjs/schemas/schemas_0.js
const _AEE = "AlreadyExistsException";
const _CME = "ConcurrentModificationException";
const _COE = "ConcurrentOperationException";
const _CR = "CreateResource";
const _CRI = "CreateResourceInput";
const _CRO = "CreateResourceOutput";
const _CRR = "CancelResourceRequest";
const _CRRI = "CancelResourceRequestInput";
const _CRRO = "CancelResourceRequestOutput";
const _CT = "ClientToken";
const _CTCE = "ClientTokenConflictException";
const _DR = "DeleteResource";
const _DRI = "DeleteResourceInput";
const _DRO = "DeleteResourceOutput";
const _DS = "DesiredState";
const _EC = "ErrorCode";
const _ET = "EventTime";
const _FM = "FailureMode";
const _GR = "GetResource";
const _GRI = "GetResourceInput";
const _GRO = "GetResourceOutput";
const _GRRS = "GetResourceRequestStatus";
const _GRRSI = "GetResourceRequestStatusInput";
const _GRRSO = "GetResourceRequestStatusOutput";
const _GSE = "GeneralServiceException";
const _HET = "HookEventTime";
const _HFE = "HandlerFailureException";
const _HIFE = "HandlerInternalFailureException";
const _HPE = "HooksProgressEvent";
const _HPEo = "HookProgressEvent";
const _HRT = "HooksRequestToken";
const _HS = "HookStatus";
const _HSM = "HookStatusMessage";
const _HTA = "HookTypeArn";
const _HTN = "HookTypeName";
const _HTVI = "HookTypeVersionId";
const _I = "Identifier";
const _ICE = "InvalidCredentialsException";
const _IP = "InvocationPoint";
const _IRE = "InvalidRequestException";
const _LR = "ListResources";
const _LRI = "ListResourcesInput";
const _LRO = "ListResourcesOutput";
const _LRR = "ListResourceRequests";
const _LRRI = "ListResourceRequestsInput";
const _LRRO = "ListResourceRequestsOutput";
const _M = "Message";
const _MR = "MaxResults";
const _NFE = "NetworkFailureException";
const _NSE = "NotStabilizedException";
const _NT = "NextToken";
const _NUE = "NotUpdatableException";
const _O = "Operation";
const _OS = "OperationStatus";
const _OSp = "OperationStatuses";
const _Op = "Operations";
const _P = "Properties";
const _PD = "PatchDocument";
const _PE = "ProgressEvent";
const _PTE = "PrivateTypeException";
const _RA = "RoleArn";
const _RAe = "RetryAfter";
const _RCE = "ResourceConflictException";
const _RD = "ResourceDescription";
const _RDe = "ResourceDescriptions";
const _RM = "ResourceModel";
const _RNFE = "ResourceNotFoundException";
const _RRSF = "ResourceRequestStatusFilter";
const _RRSS = "ResourceRequestStatusSummaries";
const _RT = "RequestToken";
const _RTNFE = "RequestTokenNotFoundException";
const _SIEE = "ServiceInternalErrorException";
const _SLEE = "ServiceLimitExceededException";
const _SM = "StatusMessage";
const _TE = "ThrottlingException";
const _TN = "TypeName";
const _TNFE = "TypeNotFoundException";
const _TVI = "TypeVersionId";
const _UAE = "UnsupportedActionException";
const _UR = "UpdateResource";
const _URI = "UpdateResourceInput";
const _URO = "UpdateResourceOutput";
const _aQE = "awsQueryError";
const _c = "client";
const _e = "error";
const _hE = "httpError";
const _s = "smithy.ts.sdk.synthetic.com.amazonaws.cloudcontrol";
const _se = "server";
const n0 = "com.amazonaws.cloudcontrol";
const { TypeRegistry } = require("@smithy/core/schema");
const { CloudControlServiceException } = require("../models/CloudControlServiceException");
const { AlreadyExistsException, ClientTokenConflictException, ConcurrentModificationException, ConcurrentOperationException, GeneralServiceException, HandlerFailureException, HandlerInternalFailureException, InvalidCredentialsException, InvalidRequestException, NetworkFailureException, NotStabilizedException, NotUpdatableException, PrivateTypeException, RequestTokenNotFoundException, ResourceConflictException, ResourceNotFoundException, ServiceInternalErrorException, ServiceLimitExceededException, ThrottlingException, TypeNotFoundException, UnsupportedActionException } = require("../models/errors");
const _s_registry = TypeRegistry.for(_s);
const CloudControlServiceException$ = [-3, _s, "CloudControlServiceException", 0, [], []];
exports.CloudControlServiceException$ = CloudControlServiceException$;
_s_registry.registerError(CloudControlServiceException$, CloudControlServiceException);
const n0_registry = TypeRegistry.for(n0);
const AlreadyExistsException$ = [-3, n0, _AEE,
    { [_aQE]: [`AlreadyExistsException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.AlreadyExistsException$ = AlreadyExistsException$;
n0_registry.registerError(AlreadyExistsException$, AlreadyExistsException);
const ClientTokenConflictException$ = [-3, n0, _CTCE,
    { [_aQE]: [`ClientTokenConflictException`, 409], [_e]: _c, [_hE]: 409 },
    [_M],
    [0]
];
exports.ClientTokenConflictException$ = ClientTokenConflictException$;
n0_registry.registerError(ClientTokenConflictException$, ClientTokenConflictException);
const ConcurrentModificationException$ = [-3, n0, _CME,
    { [_aQE]: [`ConcurrentModificationException`, 500], [_e]: _se, [_hE]: 500 },
    [_M],
    [0]
];
exports.ConcurrentModificationException$ = ConcurrentModificationException$;
n0_registry.registerError(ConcurrentModificationException$, ConcurrentModificationException);
const ConcurrentOperationException$ = [-3, n0, _COE,
    { [_aQE]: [`ConcurrentOperationException`, 409], [_e]: _c, [_hE]: 409 },
    [_M],
    [0]
];
exports.ConcurrentOperationException$ = ConcurrentOperationException$;
n0_registry.registerError(ConcurrentOperationException$, ConcurrentOperationException);
const GeneralServiceException$ = [-3, n0, _GSE,
    { [_aQE]: [`GeneralServiceException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.GeneralServiceException$ = GeneralServiceException$;
n0_registry.registerError(GeneralServiceException$, GeneralServiceException);
const HandlerFailureException$ = [-3, n0, _HFE,
    { [_aQE]: [`HandlerFailureException`, 502], [_e]: _se, [_hE]: 502 },
    [_M],
    [0]
];
exports.HandlerFailureException$ = HandlerFailureException$;
n0_registry.registerError(HandlerFailureException$, HandlerFailureException);
const HandlerInternalFailureException$ = [-3, n0, _HIFE,
    { [_aQE]: [`HandlerInternalFailureException`, 502], [_e]: _se, [_hE]: 502 },
    [_M],
    [0]
];
exports.HandlerInternalFailureException$ = HandlerInternalFailureException$;
n0_registry.registerError(HandlerInternalFailureException$, HandlerInternalFailureException);
const InvalidCredentialsException$ = [-3, n0, _ICE,
    { [_aQE]: [`InvalidCredentialsException`, 401], [_e]: _c, [_hE]: 401 },
    [_M],
    [0]
];
exports.InvalidCredentialsException$ = InvalidCredentialsException$;
n0_registry.registerError(InvalidCredentialsException$, InvalidCredentialsException);
const InvalidRequestException$ = [-3, n0, _IRE,
    { [_aQE]: [`InvalidRequestException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.InvalidRequestException$ = InvalidRequestException$;
n0_registry.registerError(InvalidRequestException$, InvalidRequestException);
const NetworkFailureException$ = [-3, n0, _NFE,
    { [_aQE]: [`NetworkFailureException`, 502], [_e]: _se, [_hE]: 502 },
    [_M],
    [0]
];
exports.NetworkFailureException$ = NetworkFailureException$;
n0_registry.registerError(NetworkFailureException$, NetworkFailureException);
const NotStabilizedException$ = [-3, n0, _NSE,
    { [_aQE]: [`NotStabilizedException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.NotStabilizedException$ = NotStabilizedException$;
n0_registry.registerError(NotStabilizedException$, NotStabilizedException);
const NotUpdatableException$ = [-3, n0, _NUE,
    { [_aQE]: [`NotUpdatableException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.NotUpdatableException$ = NotUpdatableException$;
n0_registry.registerError(NotUpdatableException$, NotUpdatableException);
const PrivateTypeException$ = [-3, n0, _PTE,
    { [_aQE]: [`PrivateTypeException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.PrivateTypeException$ = PrivateTypeException$;
n0_registry.registerError(PrivateTypeException$, PrivateTypeException);
const RequestTokenNotFoundException$ = [-3, n0, _RTNFE,
    { [_aQE]: [`RequestTokenNotFoundException`, 404], [_e]: _c, [_hE]: 404 },
    [_M],
    [0]
];
exports.RequestTokenNotFoundException$ = RequestTokenNotFoundException$;
n0_registry.registerError(RequestTokenNotFoundException$, RequestTokenNotFoundException);
const ResourceConflictException$ = [-3, n0, _RCE,
    { [_aQE]: [`ResourceConflictException`, 409], [_e]: _c, [_hE]: 409 },
    [_M],
    [0]
];
exports.ResourceConflictException$ = ResourceConflictException$;
n0_registry.registerError(ResourceConflictException$, ResourceConflictException);
const ResourceNotFoundException$ = [-3, n0, _RNFE,
    { [_aQE]: [`ResourceNotFoundException`, 404], [_e]: _c, [_hE]: 404 },
    [_M],
    [0]
];
exports.ResourceNotFoundException$ = ResourceNotFoundException$;
n0_registry.registerError(ResourceNotFoundException$, ResourceNotFoundException);
const ServiceInternalErrorException$ = [-3, n0, _SIEE,
    { [_aQE]: [`ServiceInternalErrorException`, 502], [_e]: _se, [_hE]: 502 },
    [_M],
    [0]
];
exports.ServiceInternalErrorException$ = ServiceInternalErrorException$;
n0_registry.registerError(ServiceInternalErrorException$, ServiceInternalErrorException);
const ServiceLimitExceededException$ = [-3, n0, _SLEE,
    { [_aQE]: [`ServiceLimitExceededException`, 400], [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
exports.ServiceLimitExceededException$ = ServiceLimitExceededException$;
n0_registry.registerError(ServiceLimitExceededException$, ServiceLimitExceededException);
const ThrottlingException$ = [-3, n0, _TE,
    { [_aQE]: [`ThrottlingException`, 429], [_e]: _c, [_hE]: 429 },
    [_M],
    [0]
];
exports.ThrottlingException$ = ThrottlingException$;
n0_registry.registerError(ThrottlingException$, ThrottlingException);
const TypeNotFoundException$ = [-3, n0, _TNFE,
    { [_aQE]: [`TypeNotFoundException`, 404], [_e]: _c, [_hE]: 404 },
    [_M],
    [0]
];
exports.TypeNotFoundException$ = TypeNotFoundException$;
n0_registry.registerError(TypeNotFoundException$, TypeNotFoundException);
const UnsupportedActionException$ = [-3, n0, _UAE,
    { [_aQE]: [`UnsupportedActionException`, 405], [_e]: _c, [_hE]: 405 },
    [_M],
    [0]
];
exports.UnsupportedActionException$ = UnsupportedActionException$;
n0_registry.registerError(UnsupportedActionException$, UnsupportedActionException);
exports.errorTypeRegistries = [
    _s_registry,
    n0_registry,
];
var PatchDocument = [0, n0, _PD, 8, 0];
var Properties = [0, n0, _P, 8, 0];
const CancelResourceRequestInput$ = [3, n0, _CRRI,
    0,
    [_RT],
    [0], 1
];
exports.CancelResourceRequestInput$ = CancelResourceRequestInput$;
const CancelResourceRequestOutput$ = [3, n0, _CRRO,
    0,
    [_PE],
    [[() => ProgressEvent$, 0]]
];
exports.CancelResourceRequestOutput$ = CancelResourceRequestOutput$;
const CreateResourceInput$ = [3, n0, _CRI,
    0,
    [_TN, _DS, _TVI, _RA, _CT],
    [0, [() => Properties, 0], 0, 0, [0, 4]], 2
];
exports.CreateResourceInput$ = CreateResourceInput$;
const CreateResourceOutput$ = [3, n0, _CRO,
    0,
    [_PE],
    [[() => ProgressEvent$, 0]]
];
exports.CreateResourceOutput$ = CreateResourceOutput$;
const DeleteResourceInput$ = [3, n0, _DRI,
    0,
    [_TN, _I, _TVI, _RA, _CT],
    [0, 0, 0, 0, [0, 4]], 2
];
exports.DeleteResourceInput$ = DeleteResourceInput$;
const DeleteResourceOutput$ = [3, n0, _DRO,
    0,
    [_PE],
    [[() => ProgressEvent$, 0]]
];
exports.DeleteResourceOutput$ = DeleteResourceOutput$;
const GetResourceInput$ = [3, n0, _GRI,
    0,
    [_TN, _I, _TVI, _RA],
    [0, 0, 0, 0], 2
];
exports.GetResourceInput$ = GetResourceInput$;
const GetResourceOutput$ = [3, n0, _GRO,
    0,
    [_TN, _RD],
    [0, [() => ResourceDescription$, 0]]
];
exports.GetResourceOutput$ = GetResourceOutput$;
const GetResourceRequestStatusInput$ = [3, n0, _GRRSI,
    0,
    [_RT],
    [0], 1
];
exports.GetResourceRequestStatusInput$ = GetResourceRequestStatusInput$;
const GetResourceRequestStatusOutput$ = [3, n0, _GRRSO,
    0,
    [_PE, _HPE],
    [[() => ProgressEvent$, 0], () => HooksProgressEvent]
];
exports.GetResourceRequestStatusOutput$ = GetResourceRequestStatusOutput$;
const HookProgressEvent$ = [3, n0, _HPEo,
    0,
    [_HTN, _HTVI, _HTA, _IP, _HS, _HET, _HSM, _FM],
    [0, 0, 0, 0, 0, 4, 0, 0]
];
exports.HookProgressEvent$ = HookProgressEvent$;
const ListResourceRequestsInput$ = [3, n0, _LRRI,
    0,
    [_MR, _NT, _RRSF],
    [1, 0, () => ResourceRequestStatusFilter$]
];
exports.ListResourceRequestsInput$ = ListResourceRequestsInput$;
const ListResourceRequestsOutput$ = [3, n0, _LRRO,
    0,
    [_RRSS, _NT],
    [[() => ResourceRequestStatusSummaries, 0], 0]
];
exports.ListResourceRequestsOutput$ = ListResourceRequestsOutput$;
const ListResourcesInput$ = [3, n0, _LRI,
    0,
    [_TN, _TVI, _RA, _NT, _MR, _RM],
    [0, 0, 0, 0, 1, [() => Properties, 0]], 1
];
exports.ListResourcesInput$ = ListResourcesInput$;
const ListResourcesOutput$ = [3, n0, _LRO,
    0,
    [_TN, _RDe, _NT],
    [0, [() => ResourceDescriptions, 0], 0]
];
exports.ListResourcesOutput$ = ListResourcesOutput$;
const ProgressEvent$ = [3, n0, _PE,
    0,
    [_TN, _I, _RT, _HRT, _O, _OS, _ET, _RM, _SM, _EC, _RAe],
    [0, 0, 0, 0, 0, 0, 4, [() => Properties, 0], 0, 0, 4]
];
exports.ProgressEvent$ = ProgressEvent$;
const ResourceDescription$ = [3, n0, _RD,
    0,
    [_I, _P],
    [0, [() => Properties, 0]]
];
exports.ResourceDescription$ = ResourceDescription$;
const ResourceRequestStatusFilter$ = [3, n0, _RRSF,
    0,
    [_Op, _OSp],
    [64 | 0, 64 | 0]
];
exports.ResourceRequestStatusFilter$ = ResourceRequestStatusFilter$;
const UpdateResourceInput$ = [3, n0, _URI,
    0,
    [_TN, _I, _PD, _TVI, _RA, _CT],
    [0, 0, [() => PatchDocument, 0], 0, 0, [0, 4]], 3
];
exports.UpdateResourceInput$ = UpdateResourceInput$;
const UpdateResourceOutput$ = [3, n0, _URO,
    0,
    [_PE],
    [[() => ProgressEvent$, 0]]
];
exports.UpdateResourceOutput$ = UpdateResourceOutput$;
var HooksProgressEvent = [1, n0, _HPE,
    0, () => HookProgressEvent$
];
var Operations = 64 | 0;
var OperationStatuses = 64 | 0;
var ResourceDescriptions = [1, n0, _RDe,
    0, [() => ResourceDescription$,
        0]
];
var ResourceRequestStatusSummaries = [1, n0, _RRSS,
    0, [() => ProgressEvent$,
        0]
];
exports.CancelResourceRequest$ = [9, n0, _CRR,
    2, () => CancelResourceRequestInput$, () => CancelResourceRequestOutput$
];
exports.CreateResource$ = [9, n0, _CR,
    0, () => CreateResourceInput$, () => CreateResourceOutput$
];
exports.DeleteResource$ = [9, n0, _DR,
    0, () => DeleteResourceInput$, () => DeleteResourceOutput$
];
exports.GetResource$ = [9, n0, _GR,
    0, () => GetResourceInput$, () => GetResourceOutput$
];
exports.GetResourceRequestStatus$ = [9, n0, _GRRS,
    0, () => GetResourceRequestStatusInput$, () => GetResourceRequestStatusOutput$
];
exports.ListResourceRequests$ = [9, n0, _LRR,
    0, () => ListResourceRequestsInput$, () => ListResourceRequestsOutput$
];
exports.ListResources$ = [9, n0, _LR,
    0, () => ListResourcesInput$, () => ListResourcesOutput$
];
exports.UpdateResource$ = [9, n0, _UR,
    0, () => UpdateResourceInput$, () => UpdateResourceOutput$
];
