// @aws-sdk/client-cloudcontrol@3.1069.0/dist-cjs/index.js
var __exportStar = (m, e) => { Object.assign(e, m); };
const { getAwsRegionExtensionConfiguration, resolveAwsRegionExtensionConfiguration, resolveUserAgentConfig, resolveHostHeaderConfig, getUserAgentPlugin, getHostHeaderPlugin, getLoggerPlugin, getRecursionDetectionPlugin } = require("@aws-sdk/core/client");
const { getHttpAuthSchemeEndpointRuleSetPlugin, DefaultIdentityProviderConfig, getHttpSigningPlugin, createPaginator } = require("@smithy/core");
const { getDefaultExtensionConfiguration, resolveDefaultRuntimeConfig, Client, Command, createWaiter, checkExceptions, WaiterState, createAggregatedClient } = require("@smithy/core/client");
exports.$Command = Command;
exports.__Client = Client;
const { resolveRegionConfig } = require("@smithy/core/config");
const { resolveEndpointConfig, getEndpointPlugin } = require("@smithy/core/endpoints");
const { getHttpHandlerExtensionConfiguration, resolveHttpHandlerRuntimeConfig, getContentLengthPlugin } = require("@smithy/core/protocols");
const { resolveRetryConfig, getRetryPlugin } = require("@smithy/core/retry");
const { getSchemaSerdePlugin } = require("@smithy/core/schema");
const { resolveHttpAuthSchemeConfig, defaultCloudControlHttpAuthSchemeParametersProvider } = require("./auth/httpAuthSchemeProvider");
const { getRuntimeConfig } = require("./runtimeConfig");
const { CancelResourceRequest$, CreateResource$, DeleteResource$, GetResource$, GetResourceRequestStatus$, ListResourceRequests$, ListResources$, UpdateResource$ } = require("./schemas/schemas_0");
__exportStar(require("./schemas/schemas_0"), exports);
__exportStar(require("./models/errors"), exports);
const { CloudControlServiceException } = require("./models/CloudControlServiceException");
exports.CloudControlServiceException = CloudControlServiceException;

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "cloudcontrolapi",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(getAwsRegionExtensionConfiguration(runtimeConfig), getDefaultExtensionConfiguration(runtimeConfig), getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, resolveAwsRegionExtensionConfiguration(extensionConfiguration), resolveDefaultRuntimeConfig(extensionConfiguration), resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class CloudControlClient extends Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = resolveUserAgentConfig(_config_1);
        const _config_3 = resolveRetryConfig(_config_2);
        const _config_4 = resolveRegionConfig(_config_3);
        const _config_5 = resolveHostHeaderConfig(_config_4);
        const _config_6 = resolveEndpointConfig(_config_5);
        const _config_7 = resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(getUserAgentPlugin(this.config));
        this.middlewareStack.use(getRetryPlugin(this.config));
        this.middlewareStack.use(getContentLengthPlugin(this.config));
        this.middlewareStack.use(getHostHeaderPlugin(this.config));
        this.middlewareStack.use(getLoggerPlugin(this.config));
        this.middlewareStack.use(getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: defaultCloudControlHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class CancelResourceRequestCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "CancelResourceRequest", {})
    .n("CloudControlClient", "CancelResourceRequestCommand")
    .sc(CancelResourceRequest$)
    .build() {
}

class CreateResourceCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "CreateResource", {})
    .n("CloudControlClient", "CreateResourceCommand")
    .sc(CreateResource$)
    .build() {
}

class DeleteResourceCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "DeleteResource", {})
    .n("CloudControlClient", "DeleteResourceCommand")
    .sc(DeleteResource$)
    .build() {
}

class GetResourceCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "GetResource", {})
    .n("CloudControlClient", "GetResourceCommand")
    .sc(GetResource$)
    .build() {
}

class GetResourceRequestStatusCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "GetResourceRequestStatus", {})
    .n("CloudControlClient", "GetResourceRequestStatusCommand")
    .sc(GetResourceRequestStatus$)
    .build() {
}

class ListResourceRequestsCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "ListResourceRequests", {})
    .n("CloudControlClient", "ListResourceRequestsCommand")
    .sc(ListResourceRequests$)
    .build() {
}

class ListResourcesCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "ListResources", {})
    .n("CloudControlClient", "ListResourcesCommand")
    .sc(ListResources$)
    .build() {
}

class UpdateResourceCommand extends Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CloudApiService", "UpdateResource", {})
    .n("CloudControlClient", "UpdateResourceCommand")
    .sc(UpdateResource$)
    .build() {
}

const paginateListResourceRequests = createPaginator(CloudControlClient, ListResourceRequestsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListResources = createPaginator(CloudControlClient, ListResourcesCommand, "NextToken", "NextToken", "MaxResults");

const checkState = async (client, input) => {
    let reason;
    try {
        let result = await client.send(new GetResourceRequestStatusCommand(input));
        reason = result;
        try {
            const returnComparator = () => {
                return result.ProgressEvent.OperationStatus;
            };
            if (returnComparator() === "SUCCESS") {
                return { state: WaiterState.SUCCESS, reason };
            }
        }
        catch (e) { }
        try {
            const returnComparator = () => {
                return result.ProgressEvent.OperationStatus;
            };
            if (returnComparator() === "FAILED") {
                return { state: WaiterState.FAILURE, reason };
            }
        }
        catch (e) { }
        try {
            const returnComparator = () => {
                return result.ProgressEvent.OperationStatus;
            };
            if (returnComparator() === "CANCEL_COMPLETE") {
                return { state: WaiterState.FAILURE, reason };
            }
        }
        catch (e) { }
    }
    catch (exception) {
        reason = exception;
    }
    return { state: WaiterState.RETRY, reason };
};
const waitForResourceRequestSuccess = async (params, input) => {
    const serviceDefaults = { minDelay: 5, maxDelay: 120 };
    return createWaiter({ ...serviceDefaults, ...params }, input, checkState);
};
const waitUntilResourceRequestSuccess = async (params, input) => {
    const serviceDefaults = { minDelay: 5, maxDelay: 120 };
    const result = await createWaiter({ ...serviceDefaults, ...params }, input, checkState);
    return checkExceptions(result);
};

const commands = {
    CancelResourceRequestCommand,
    CreateResourceCommand,
    DeleteResourceCommand,
    GetResourceCommand,
    GetResourceRequestStatusCommand,
    ListResourceRequestsCommand,
    ListResourcesCommand,
    UpdateResourceCommand,
};
const paginators = {
    paginateListResourceRequests,
    paginateListResources,
};
const waiters = {
    waitUntilResourceRequestSuccess,
};
class CloudControl extends CloudControlClient {
}
createAggregatedClient(commands, CloudControl, { paginators, waiters });

const HandlerErrorCode = {
    ACCESS_DENIED: "AccessDenied",
    ALREADY_EXISTS: "AlreadyExists",
    GENERAL_SERVICE_EXCEPTION: "GeneralServiceException",
    INTERNAL_FAILURE: "InternalFailure",
    INVALID_CREDENTIALS: "InvalidCredentials",
    INVALID_REQUEST: "InvalidRequest",
    NETWORK_FAILURE: "NetworkFailure",
    NOT_FOUND: "NotFound",
    NOT_STABILIZED: "NotStabilized",
    NOT_UPDATABLE: "NotUpdatable",
    RESOURCE_CONFLICT: "ResourceConflict",
    SERVICE_INTERNAL_ERROR: "ServiceInternalError",
    SERVICE_LIMIT_EXCEEDED: "ServiceLimitExceeded",
    SERVICE_TIMEOUT: "ServiceTimeout",
    THROTTLING: "Throttling",
    UNAUTHORIZED_TAGGING_OPERATION: "UnauthorizedTaggingOperation",
};
const Operation = {
    CREATE: "CREATE",
    DELETE: "DELETE",
    UPDATE: "UPDATE",
};
const OperationStatus = {
    CANCEL_COMPLETE: "CANCEL_COMPLETE",
    CANCEL_IN_PROGRESS: "CANCEL_IN_PROGRESS",
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    PENDING: "PENDING",
    SUCCESS: "SUCCESS",
};

exports.CancelResourceRequestCommand = CancelResourceRequestCommand;
exports.CloudControl = CloudControl;
exports.CloudControlClient = CloudControlClient;
exports.CreateResourceCommand = CreateResourceCommand;
exports.DeleteResourceCommand = DeleteResourceCommand;
exports.GetResourceCommand = GetResourceCommand;
exports.GetResourceRequestStatusCommand = GetResourceRequestStatusCommand;
exports.HandlerErrorCode = HandlerErrorCode;
exports.ListResourceRequestsCommand = ListResourceRequestsCommand;
exports.ListResourcesCommand = ListResourcesCommand;
exports.Operation = Operation;
exports.OperationStatus = OperationStatus;
exports.UpdateResourceCommand = UpdateResourceCommand;
exports.paginateListResourceRequests = paginateListResourceRequests;
exports.paginateListResources = paginateListResources;
exports.waitForResourceRequestSuccess = waitForResourceRequestSuccess;
exports.waitUntilResourceRequestSuccess = waitUntilResourceRequestSuccess;
