// aws-sdk-client-mock-vitest@4.0.1/dist/matcher.js
import { ObjectContaining } from '@vitest/expect';
import { notNull, ordinalOf } from './utils.js';
function formatCalls(context, client, command, expectedCall, message) {
    const calls = client.commandCalls(command);
    return calls.length === 0
        ? message
        : [
            message,
            '',
            'Received:',
            '',
            ...calls.flatMap((call, index) => {
                /* eslint-disable @typescript-eslint/no-unsafe-assignment */
                const input = call.args[0].input;
                return [
                    `   ${ordinalOf(index + 1)} ${command.name} call`,
                    '',
                    expectedCall
                        ? context.utils.diff(expectedCall, input, { omitAnnotationLines: true })
                        : context.utils
                            .stringify(input)
                            .split('\n')
                            .map(line => `    ${line}`)
                            .join('\n'),
                    '',
                ].filter(notNull);
            }),
            `Number of calls: ${calls.length.toString()}`,
        ].join('\n');
}
const toHaveReceivedCommandTimes = function (client, command, times) {
    const { isNot } = this;
    const callCount = client.commandCalls(command).length;
    const pass = callCount === times;
    return {
        message: () => {
            const message = isNot
                ? `expected "${command.name}" to not be called ${times.toString()} times`
                : `expected "${command.name}" to be called ${times.toString()} times, but got ${callCount.toString()} times`;
            return formatCalls(this, client, command, undefined, message);
        },
        pass,
    };
};
const toReceiveCommandTimes = toHaveReceivedCommandTimes;
const toHaveReceivedCommandOnce = function (client, command) {
    const { isNot } = this;
    const callCount = client.commandCalls(command).length;
    const pass = callCount === 1;
    return {
        message: () => {
            const message = isNot
                ? `expected "${command.name}" to not be called once`
                : `expected "${command.name}" to be called once, but got ${callCount.toString()} times`;
            return formatCalls(this, client, command, undefined, message);
        },
        pass,
    };
};
const toReceiveCommandOnce = toHaveReceivedCommandOnce;
const toHaveReceivedCommand = function (client, command) {
    const { isNot } = this;
    const callCount = client.commandCalls(command).length;
    const pass = callCount >= 1;
    return {
        message: () => {
            const message = isNot
                ? `expected "${command.name}" to not be called at all, but actually been called ${callCount.toString()} times`
                : `expected "${command.name}" to be called at least once`;
            return formatCalls(this, client, command, undefined, message);
        },
        pass,
    };
};
const toReceiveCommand = toHaveReceivedCommand;
const toHaveReceivedCommandWith = function (client, command, input) {
    const { isNot, utils } = this;
    const calls = client.commandCalls(command);
    const pass = calls.some(call => new ObjectContaining(input).asymmetricMatch(call.args[0].input));
    return {
        message: () => {
            const message = isNot
                ? `expected "${command.name}" to not be called with arguments: ${utils.printExpected(input)}`
                : `expected "${command.name}" to be called with arguments: ${utils.printExpected(input)}`;
            return formatCalls(this, client, command, input, message);
        },
        pass,
    };
};
const toReceiveCommandWith = toHaveReceivedCommandWith;
/*

  */
const toHaveReceivedNthCommandWith = function (client, command, times, input) {
    const { isNot, utils } = this;
    const calls = client.commandCalls(command);
    const call = calls.length < times ? undefined : calls[times - 1];
    const pass = call
        ? new ObjectContaining(input).asymmetricMatch(call.args[0].input)
        : false;
    return {
        message: () => {
            const message = isNot
                ? `expected ${ordinalOf(times)} "${command.name}" to not be called with arguments: ${utils.printExpected(input)}`
                : `expected ${ordinalOf(times)} "${command.name}" to be called with arguments: ${utils.printExpected(input)}`;
            return formatCalls(this, client, command, input, message);
        },
        pass,
    };
};
const toReceiveNthCommandWith = toHaveReceivedNthCommandWith;
const toHaveReceivedLastCommandWith = function (client, command, input) {
    const { isNot, utils } = this;
    const calls = client.commandCalls(command);
    const call = calls.length === 0 ? undefined : calls[calls.length - 1];
    const pass = call
        ? new ObjectContaining(input).asymmetricMatch(call.args[0].input)
        : false;
    return {
        message: () => {
            const message = isNot
                ? `expected last "${command.name}" to not be called with arguments: ${utils.printExpected(input)}`
                : `expected last "${command.name}" to be called with arguments: ${utils.printExpected(input)}`;
            return formatCalls(this, client, command, input, message);
        },
        pass,
    };
};
const toReceiveLastCommandWith = toHaveReceivedLastCommandWith;
export { toHaveReceivedCommand, toHaveReceivedCommandOnce, toHaveReceivedCommandTimes, toHaveReceivedCommandWith, toHaveReceivedLastCommandWith, toHaveReceivedNthCommandWith, toReceiveCommand, toReceiveCommandOnce, toReceiveCommandTimes, toReceiveCommandWith, toReceiveLastCommandWith, toReceiveNthCommandWith, };
